drivers: i2c: Add Infineon CAT1 i2c driver

sreeramIfx opened the pull request:
- Add initial version of Infineon CAT1 i2c driver.
 - Add initial version of binding file for Infineon  CAT1 I2C driver

npal-cy commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,539 @@
+/**
+ * @brief I2C driver for Infineon CAT1 MCU family.
+ *
+ * Note:
```
looks like it is out of date note. Slave mode is supported now. Please remove this note.

npal-cy commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,539 @@
+	 CYHAL_I2C_SLAVE_ERR_EVENT)
+
+#if !defined(IFX_CAT1_I2C_DO_NOT_USE_ASYNC)
+#define IFX_CAT1_I2C_ASYNC
```
Lets move IFX_CAT1_I2C_ASYNC in Kconfig (with default y)

npal-cy commented on drivers/i2c/Kconfig.ifx_cat1:
```diff
@@ -0,0 +1,20 @@
+# Infineon CAT1 I2C configuration options
+
+# Copyright (c) 2022 Cypress Semiconductor Corporation.
```
2023

npal-cy commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,539 @@
+	return 0;
+}
+
+static int ifx_i2c_infineon_cat1_msg_validate(struct i2c_msg *msg, uint8_t num_msgs)
```
ifx_i2c_infineon_cat1_msg_validate -> ifx_cat1_i2c_msg_validate

sreeramIfx commented on drivers/i2c/Kconfig.ifx_cat1:
```diff
@@ -0,0 +1,20 @@
+	help
+	  This option enables the I2C driver for Infineon CAT1 family.
+
+config I2C_INFINEON_CAT1_SLAVE_BUF
```
updated to use target instead of slave..

ifyall commented on dts/arm/infineon/psoc6/psoc6_02/psoc6_02.dtsi:
```diff
@@ -9,6 +9,110 @@
+			interrupts = <49 6>;
+			status = "disabled";
+		};
+		i2c12: i2c@406c0000 {
```
There are 13 **blocks** on this part that can be I2C. They could also be a UART or SPI, if the user desires. They are called Serial Communication Blocks (SCB).

teburd replied:
Understood! I believe NXP's FlexComm/FlexIO, and RPI's PIO are similar in that regard and there is a bit of a discussion around those. It might be worth reviewing how those drivers and soc device tree definitions are laid out.

There is for example i2c_mcux_flexcomm and a ton of flexcomms on the rt685

The dtsi for that part is here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rt6xx_common.dtsi

And the flexcomms get used as i2c for example here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts#L161

So you can split up the soc dtsi that defines your serial comm blocks and then instantiations of those comm blocks as i2c buses to the board devicetree.

ifyall replied:
@teburd, thanks for the pointers. Given that we already have the UART version of this up with the pattern you highlighted, can we take this as an issue to fix in a later PR? We want to do the review and design of the NXP mechanism and how to apply it to our SoCs as its own task.

teburd replied:
That's fine, but I'd just reiterate doing this sooner rather than later is probably a good idea.

sreeramIfx commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,531 @@
+	}
+
+	/* Acquire semaphore (block I2C transfer for another thread) */
+	ret = k_sem_take(&data->operation_sem, K_MSEC(I2C_CAT1_SEM_TIMEOUT_MS));
```
I have updated the code to use K_FOREVER

sreeramIfx commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,531 @@
+	}
+
+	/* This function checks if msg.buf is not NULL and if
+	 * slave address is not 10 bit.
```
updated..

sreeramIfx commented on tests/drivers/i2c/i2c_api/prj.conf:
```diff
@@ -1,3 +1,4 @@
 CONFIG_I2C=y
 CONFIG_ZTEST=y
 CONFIG_ZTEST_NEW_API=y
+CONFIG_HEAP_MEM_POOL_SIZE=1024
```
this was added since the driver was using malloc, removed it now since the driver is updated to not use malloc and allocate statically

sreeramIfx commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,531 @@
+	}
+
+	data->p_target_wr_buffer = NULL;
+	data->p_target_wr_buffer = k_malloc(CONFIG_I2C_INFINEON_CAT1_TARGET_BUF);
```
updated the driver to allocate this buffer statically..

teburd commented on drivers/i2c/i2c_ifx_cat1.c:
```diff
@@ -0,0 +1,519 @@
+                                                                                                   \
+	I2C_DEVICE_DT_INST_DEFINE(n, ifx_cat1_i2c_init, NULL, &ifx_cat1_i2c_data##n,               \
+				  &i2c_cat1_cfg_##n, POST_KERNEL,                                  \
+				  CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &i2c_cat1_driver_api);
```
```suggestion
				  CONFIG_I2C_INIT_PRIORITY, &i2c_cat1_driver_api);
```

sreeramIfx committed:
drivers: i2c: Add Infineon CAT1 i2c driver

 - Add initial version of Infineon CAT1 i2c driver.
 - Add initial version of binding file for Infineon
   CAT1 I2C driver

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>
